### PR TITLE
Add discord rpc, add speed to draw fps, add godmode

### PR DESCRIPTION
**Pull request by fedddddd**

### CODE REVIEW

**Comment by RektInator** [comment]
I don't think the source code of the discord rpc belongs in the client, you should use a submodule and create a script for premake.

**Comment by RektInator** [comment]
The god command should just set the godmode flag on the entity, changing the max health is not godmode

**Comment by RektInator** [comment]
I also think PRs should only have a single purpose, so 1 PR per bug fix / feature.

**Comment by fedddddd** [comment]
Couldn't find the flag for godmode so i just did that but yea

File: src/client/module/discord_rpc.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+#include <string.h>
+
+namespace 
+{
+	discord::Core* core{};
+	auto result = discord::Core::Create(757965986954477690, DiscordCreateFlags_Default, &core);
+	discord::Activity activity{};
+
+	int get_player_count() {
+		int maxclients = game::native::Dvar_FindVar("sv_maxclients")->current.integer;
```
Comment (RektInator) [inline src/client/module/discord_rpc.cpp:21]: auto

File: src/client/module/discord_rpc.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		int count = 0;
```
Comment (RektInator) [inline src/client/module/discord_rpc.cpp:22]: auto

File: src/client/module/discord_rpc.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		for (int i = 0; i < maxclients; i++) {
```
Comment (RektInator) [inline src/client/module/discord_rpc.cpp:23]: auto, bracket on new line

File: src/client/module/discord_rpc.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+#include "game/game.hpp"
+#include "game/dvars.hpp"
+#include "discord/discord.h"
+
+#include "utils/hook.hpp"
+#include "utils/string.hpp"
+#include <string.h>
+
+namespace
```
Comment (RektInator) [inline src/client/module/discord_rpc.cpp:14]: you could make this code part of the class defined below

File: src/client/module/command.cpp [hunk 5]
```diff
@@ -169,6 +169,12 @@ void command::add_mp_commands()
         game::native::mp::g_entities[clientNum].client->ps.origin[2] = std::strtof(params.get(3), NULL);
     });
 
+    command::add_sv("god", [&](int clientNum, command::params_sv&) {
+        game::native::mp::g_entities[clientNum].maxHealth = game::native::mp::g_entities[clientNum].maxHealth == INT32_MAX ? 100 : INT32_MAX;
```
Comment (RektInator) [inline src/client/module/command.cpp:173]: change the godmode flag instead

File: src/client/module/command.cpp [hunk 6]
```diff
@@ -169,6 +169,12 @@ void command::add_mp_commands()
         game::native::mp::g_entities[clientNum].client->ps.origin[2] = std::strtof(params.get(3), NULL);
     });
 
+    command::add_sv("god", [&](int clientNum, command::params_sv&) {
```
Comment (RektInator) [inline src/client/module/command.cpp:172]: bracket on next line

**Review by RektInator — CHANGES_REQUESTED** [review]

File: src/client/module/fps.cpp [hunk 7]
```diff
@@ -104,9 +105,19 @@ namespace
 			if (game::native::mp::g_entities && game::native::Dvar_FindVar("cg_drawFPS")->current.integer > 1 && game::native::SV_Loaded())
 			{
 				auto* const origin_string = utils::string::va("%f, %f, %f", game::native::mp::g_entities[0].client->ps.origin[0], game::native::mp::g_entities[0].client->ps.origin[1], game::native::mp::g_entities[0].client->ps.origin[2]);
+				
 				const auto origin_x = (game::native::ScrPlace_GetViewPlacement()->realViewportSize[0] - 10.0f) - game::native::R_TextWidth(origin_string, 0x7FFFFFFF, font) * scale;
+				
 				game::native::R_AddCmdDrawText(origin_string, 0x7FFFFFFF, font, origin_x, y + 50, scale, scale, 0.0f, origin_color, 6);
 			}
+			if (game::native::mp::g_entities && game::native::Dvar_FindVar("cg_drawFPS")->current.integer > 2 && game::native::SV_Loaded()) {
```
Comment (momo5502) [inline src/client/module/fps.cpp:113]: empty line

File: src/client/module/discord_rpc.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+
+class discord_rpc final : public module
+{
+public:
+	void post_unpack() override
+	{
+		activity.GetTimestamps().SetStart(std::time(0));
+		scheduler::loop(tick, scheduler::pipeline::renderer);
```
Comment (momo5502) [inline src/client/module/discord_rpc.cpp:67]: Renderer??? Really???

File: src/client/module/discord_rpc.cpp (hunk 4 above)
Comment (momo5502) [reply src/client/module/discord_rpc.cpp:14]: no, anonymous namespaces are better

File: src/client/module/discord_rpc.cpp [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+#include "discord/discord.h"
+
+#include "utils/hook.hpp"
+#include "utils/string.hpp"
+#include <string.h>
+
+namespace 
+{
+	discord::Core* core{};
+	auto result = discord::Core::Create(757965986954477690, DiscordCreateFlags_Default, &core);
```
Comment (momo5502) [inline src/client/module/discord_rpc.cpp:17]: What's this id?

File: src/client/module/discord_rpc.cpp [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+	void tick() {
+		if (game::is_mp())
+		{
+			if (game::native::SV_Loaded())
+			{
+				activity.GetParty().GetSize().SetCurrentSize(get_player_count());
+				activity.GetParty().GetSize().SetMaxSize(game::native::Dvar_FindVar("sv_maxclients")->current.integer);
+
+				activity.SetDetails("Multiplayer");
+				activity.SetState(utils::string::va("Playing on %s", game::native::Dvar_FindVar("mapname")->current.string));
```
Comment (momo5502) [inline src/client/module/discord_rpc.cpp:40]: Might wanna localize that?

File: src/client/module/discord_rpc.cpp [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+{
+public:
+	void post_unpack() override
+	{
+		activity.GetTimestamps().SetStart(std::time(0));
+		scheduler::loop(tick, scheduler::pipeline::renderer);
+	}
+};
+
+REGISTER_MODULE(discord_rpc);
```
Comment (momo5502) [inline src/client/module/discord_rpc.cpp:71]: Empty line?

File: src/client/module/discord_rpc.cpp [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+		int count = 0;
+		for (int i = 0; i < maxclients; i++) {
+			if (game::native::mp::g_entities[i].s.entityType) {
+				count++;
+			}
+		}
+		return count;
+	}
+
+	void tick() {
```
Comment (momo5502) [inline src/client/module/discord_rpc.cpp:31]: Bracket?

File: src/client/module/discord_rpc.cpp [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+		if (game::is_mp())
+		{
+			if (game::native::SV_Loaded())
+			{
```
Comment (momo5502) [inline src/client/module/discord_rpc.cpp:35]: That nesting level...

**Review by momo5502 — CHANGES_REQUESTED** [review]
use submodules

**Comment by momo5502** [comment]
Oh and the build pipeline? Might wanna create pull requests that actually build?

File: src/client/module/discord_rpc.cpp (hunk 4 above)
Comment (RektInator) [reply src/client/module/discord_rpc.cpp:14]: alright, wasn't up-to-date on the new code guidelines for IW6x, my bad :P

File: src/client/module/discord_rpc.cpp (hunk 9 above)
Comment (RektInator) [reply src/client/module/discord_rpc.cpp:17]: He probably generated his own ID for the game on discord.com, I'd recommend generating one ourselves.

Comment (fedddddd) [reply src/client/module/discord_rpc.cpp:17]: Discord application clientid

**Comment by fedddddd** [comment]
Yea if only discord game sdk was on github

File: deps/premake/discord-rpc.lua [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+end
+
+function discordrpc.project()
+	project "discord-rpc"
+		language "C++"
+
+		discordrpc.includes()
+
+		files {
+			path.join(discordrpc.source, "src/serialization.h"),
```
Comment (momo5502) [inline deps/premake/discord-rpc.lua:25]: You can use wildcards for that

File: deps/premake/discord-rpc.lua [hunk 15]
```diff
@@ -0,0 +1,40 @@
+discordrpc = {
+	source = path.join(dependencies.basePath, "discord-rpc"),
+	rapidjson = path.join(dependencies.basePath, "rapidjson"),
```
Comment (momo5502) [inline deps/premake/discord-rpc.lua:3]: That's not how our dependency management works

File: src/client/module/command.cpp [hunk 16]
```diff
@@ -157,6 +157,28 @@ void command::add_mp_commands()
...
     });
 
+    command::add_sv("god", [&](int clientNum, command::params_sv&)
+    {
+        if (!game::native::SV_Loaded())
+        {
+            return;
+        }
+
+        game::native::mp::g_entities[clientNum].flags ^= 1;
```
Comment (momo5502) [inline src/client/module/command.cpp:167]: I want this command SP only

**Review by momo5502 — CHANGES_REQUESTED** [review]
Please create a new pull request for the god mode. The discord integration will be done by the main developers

**Pull request closed without merging**